File: python/src/benchmarks/ycsb/populate.py
import sys
import threading
import time
from concurrent.futures import ThreadPoolExecutor, as_completed

from google.cloud.spanner_v1.database import Database

from .utils import build_key_name, generate_random_string


def populate_data(
    database: Database,
    table_name: str = "usertable",
    record_count: int = 100000,
    zero_padding: int = 12,
    field_count: int = 10,
    field_length: int = 100,
    threads: int = 16,
    batch_size: int = 500,
) -> None:
    """
    Populates the YCSB table with pseudo-random records concurrently across multiple worker threads.
    Precomputes column names once and uses batched mutations for high throughput.
    """
    if record_count <= 0:
        return

    ranges = _compute_partition_ranges(record_count, threads)
    if not ranges:
        return

    # Precompute column names once to avoid per-record string/list allocations
    col_names = ["id"] + [f"field{f}" for f in range(field_count)]

    print(
        f"Starting data population: {record_count} records across {len(ranges)} threads with batch size {batch_size}..."
    )
    sys.stdout.flush()

    start_time = time.time()
    last_log_time_box = [start_time]
    progress = 0
    progress_lock = threading.Lock()

    def worker_populate(start_idx: int, end_idx: int) -> None:
        nonlocal progress
        batch_values = []

        for record_idx in range(start_idx, end_idx):
            key = build_key_name(record_idx, zero_padding)
            row = [key] + [
                generate_random_string(field_length) for _ in range(field_count)
            ]
            batch_values.append(row)

            if len(batch_values) >= batch_size:
                with database.batch() as batch:
                    batch.insert_or_update(
                        table=table_name,
                        columns=col_names,
                        values=batch_values,
                    )
                with progress_lock:
                    progress += len(batch_values)
                    _log_progress(
                        progress,
                        record_count,
                        start_time,
                        last_log_time_box,
                    )
                batch_values = []

        if batch_values:
            with database.batch() as batch:
                batch.insert_or_update(
                    table=table_name,
                    columns=col_names,
                    values=batch_values,
                )
            with progress_lock:
                progress += len(batch_values)
                _log_progress(
                    progress,
                    record_count,
                    start_time,
                    last_log_time_box,
                )

    with ThreadPoolExecutor(max_workers=len(ranges)) as executor:
        futures = [
            executor.submit(worker_populate, start, end) for start, end in ranges
        ]
        for f in as_completed(futures):
            f.result()

    total_duration = time.time() - start_time
    total_rate = (float(record_count) / total_duration) if total_duration > 0 else 0.0
    print(
        f"Data population complete: {record_count} records inserted in {total_duration:.3f}s ({total_rate:.2f} records/sec)."
    )
    sys.stdout.flush()


def _compute_partition_ranges(
    record_count: int, num_partitions: int
) -> list[tuple[int, int]]:
    if num_partitions <= 0 or record_count <= 0:
        return []
    if num_partitions > record_count:
        num_partitions = record_count

    chunk_size = record_count // num_partitions
    remainder = record_count % num_partitions
    ranges = []
    current_start = 0

    for i in range(num_partitions):
        size = chunk_size + (1 if i < remainder else 0)
        ranges.append((current_start, current_start + size))
        current_start += size

    return ranges


def _log_progress(
    current: int,
    total: int,
    start_time: float,
    last_log_time_box: list[float],
) -> None:
    now = time.time()
    if now - last_log_time_box[0] >= 5.0 or current == total:
        last_log_time_box[0] = now
        pct = (float(current) * 100.0) / float(total) if total > 0 else 100.0
        elapsed = max(1.0, now - start_time)
        rate = int(float(current) / elapsed)
        print(f"Progress: {current} / {total} records ({pct:.1f}%) - {rate} records/s")
        sys.stdout.flush()

File: node/src/benchmarks/ycsb/populate.ts
import {Database, MutationSet} from '@google-cloud/spanner';
import {buildKeyName, buildRow} from './utils';

/**
 * Inserts initial records into Cloud Spanner in parallel using batched mutations.
 */
export async function populateData(
  database: Database,
  tableName: string,
  recordCount: number,
  zeroPadding: number,
  fieldCount: number,
  fieldLength: number,
  batchSize: number,
  threads: number,
  skipData = false,
): Promise<void> {
  if (skipData) {
    console.log('Skipping data population (--skip-data flag specified).');
    return;
  }

  const effectiveBatchSize = batchSize > 0 ? batchSize : 500;
  const effectiveThreads = threads > 0 ? threads : 16;
  const ranges = computePartitionRanges(recordCount, effectiveThreads);

  if (ranges.length === 0) {
    return;
  }

  let completedRecords = 0;
  let lastLogTime = Date.now();
  const startTime = Date.now();

  console.log(
    `Starting data population: ${recordCount} records across ${ranges.length} threads with batch size ${effectiveBatchSize}...`,
  );

  const logProgress = (count: number) => {
    const now = Date.now();
    if (now - lastLogTime >= 5000 || count === recordCount) {
      lastLogTime = now;
      const pct = ((count * 100.0) / recordCount).toFixed(1);
      const elapsedSeconds = Math.max(1, (now - startTime) / 1000);
      const rate = Math.floor(count / elapsedSeconds);
      console.log(
        `Progress: ${count} / ${recordCount} records (${pct}%) - ${rate} records/s`,
      );
    }
  };

  const workers = ranges.map(async range => {
    let batch: Array<Record<string, string>> = [];

    for (
      let recordIndex = range.start;
      recordIndex < range.end;
      recordIndex++
    ) {
      const key = buildKeyName(recordIndex, zeroPadding);
      const row = buildRow(key, fieldCount, fieldLength);
      batch.push(row);

      if (batch.length >= effectiveBatchSize) {
        const mutations = new MutationSet();
        mutations.upsert(tableName, batch);
        await database.writeAtLeastOnce(mutations);
        completedRecords += batch.length;
        logProgress(completedRecords);
        batch = [];
      }
    }

    if (batch.length > 0) {
      const mutations = new MutationSet();
      mutations.upsert(tableName, batch);
      await database.writeAtLeastOnce(mutations);
      completedRecords += batch.length;
      logProgress(completedRecords);
      batch = [];
    }
  });

  await Promise.all(workers);

  const totalDurationSeconds = Math.max(0.001, (Date.now() - startTime) / 1000);
  const finalRate = (recordCount / totalDurationSeconds).toFixed(2);
  console.log(
    `Data population complete: ${recordCount} records inserted in ${totalDurationSeconds.toFixed(3)}s (${finalRate} records/sec).`,
  );
}

interface PartitionRange {
  start: number;
  end: number;
}

function computePartitionRanges(
  recordCount: number,
  threads: number,
): PartitionRange[] {
  if (threads <= 0 || recordCount <= 0) {
    return [];
  }
  const effectiveThreads = Math.min(threads, recordCount);
  const baseChunk = Math.floor(recordCount / effectiveThreads);
  const remainder = recordCount % effectiveThreads;
  const ranges: PartitionRange[] = [];
  let currentStart = 0;

  for (let i = 0; i < effectiveThreads; i++) {
    const chunkSize = baseChunk + (i < remainder ? 1 : 0);
    ranges.push({
      start: currentStart,
      end: currentStart + chunkSize,
    });
    currentStart += chunkSize;
  }
  return ranges;
}
